### ISSUE TITLE
Consider integration in core and std ?

### ISSUE THREAD

**Issue by Urgau** [aldanor/fast-float-rust#15]
Have you consider replacing the `dec2flt` algorithm in the standard library by your crate ?

I remember have seen several members of the Rust Libs Team encouraging people how have better algorithm (fast, efficient, ...) to integrate their crate in the std library.

*Source of [dec2flt](https://github.com/rust-lang/rust/tree/master/library/core/src/num/dec2flt)*

**Comment by aldanor** [comment]
@Urgau I remember this has been already discussed somewhere in rustc GH issues and as I was searching for it, I noticed a related question has been raised again just yesterday - see https://github.com/rust-lang/rust/issues/85198 for further details.

One of concerns to keep in mind: binary size; both dec2flt and the algo used here use lookup tables, but dec2flt will be a bit leaner, I believe.

**Comment by Urgau** [comment]
Yeah I saw this issue, it's the primary reason why I search crate that do really fast float parsing and ask you that question.

> One of concerns to keep in mind: binary size; both dec2flt and the algo used here use lookup tables, but dec2flt will be a bit leaner, I believe.

I know that for certain target, like embedded object, binary size is a big deal but for vast majority of users who use Rust on computers, size is not really a big deal. I think that most people will accept +10/20 kio (just a guess not a actual number) increase for blazing performance improvement in float parsing and for size constraint environment a `cfg` with `-Zbuild-std` for disabling the fast algo and go to a slower one in exchange for binary size reduce.

If binary size is the only major concern I think you should consider making a pull request to the rust repository.

**Comment by lemire** [comment]
I think that what @Urgau wrote makes sense. The binary size can be important, but that's something one can tweak depending on one's needs.

**Comment by Alexhuszagh** [comment]
@aldanor I'm working on gradual integrations of faster float-parsing algorithms into dec2flt, although obviously, faster algorithms would be preferable if applicable (the benchmarks for this look really promising).

As detailed in this internals [post](https://internals.rust-lang.org/t/implementing-a-fast-correct-float-parser/14670), I'm looking to fix in the following order:

- Fix disguised fast-path algorithms.
- Fix the `Bellerophon` algorithm, which incorrectly requires arbitrary-precision arithmetic for the significant digits currently (making it very slow).
- Implement faster, arbitrary-precision algorithms (both use iterative scaling based on powers of 2, making them very slow).

My goal is gradual changes, since obviously, it's much easier to justify these changes with minimal differences to binary size but major overall performance improvements. Especially since @hanna-kruppe, the author of the float parsing algorithms, has [retired](https://github.com/rust-lang/team/pull/400) from Rust development, meaning people reviewing these changes are less familiar with the codebase or float-parsing algorithms in general.

Any feedback however would be greatly appreciated.

**EDIT** In conversations a few years ago with @hanna-kruppe, it was very clear that both binary size was important, and there was at the time a tepid response to moving away from `Big32x40` (which can only hold ~385 digits, which is insufficient for a lot of floating-point cases).

**Comment by Alexhuszagh** [comment]
I should add, I've got a working repository for implementing these changes into Rust's core library [here](https://github.com/Alexhuszagh/rust-dec2flt), and the different branches provide improvements to different components of dec2flt.

- [core](https://github.com/Alexhuszagh/rust-dec2flt/tree/core): A minimal, extracted version identical to Rust's dec2flt implementation.
- [disguised](https://github.com/Alexhuszagh/rust-dec2flt/tree/disguised): Handle disguised fast-path cases for better performance.
- [infer](https://github.com/Alexhuszagh/rust-dec2flt/tree/infer): Reduce binary sizes by inferring the binary exponent, rather than explicitly store them (no performance penalty).
- [moderate](https://github.com/Alexhuszagh/rust-dec2flt/tree/moderate): Fix major performance penalties in the Bellerophon algorithm.

I'll be working on slow-path algorithms later, and my goal is a quick but comprehensive fix of `dec2flt` in a way that should minimize the number of changes made initially, which hopefully should ease the burden of integrating these changes. I'm wholly amenable to making small, incremental changes and then having all improvements replaced by integration of this library, and if I can help in any way, let me know.

**Comment by lemire** [comment]
@Alexhuszagh 

One point to consider is that the current library can effectively act as a fast path. That is, there is no need to throw away anything but maybe an existing fast path. So it can be non-invasive (and also, easily made optional).

In the Go runtime, the equivalent of the code in this repo. is running. All they did was to remove an existing fast path and replace it with a new one.

It is also relatively thin. Here is the gist of the code...

```Rust

pub fn compute_float<F: Float>(q: i64, mut w: u64) -> AdjustedMantissa {
    let am_zero = AdjustedMantissa::zero_pow2(0);
    let am_inf = AdjustedMantissa::zero_pow2(F::INFINITE_POWER);
    let am_error = AdjustedMantissa::zero_pow2(-1);

    if w == 0 || q < F::SMALLEST_POWER_OF_TEN as i64 {
        return am_zero;
    } else if q > F::LARGEST_POWER_OF_TEN as i64 {
        return am_inf;
    }
    let lz = w.leading_zeros();
    w <<= lz;
    let (lo, hi) = compute_product_approx(q, w, F::MANTISSA_EXPLICIT_BITS + 3);
    if lo == 0xFFFF_FFFF_FFFF_FFFF {
        let inside_safe_exponent = (q >= -27) && (q <= 55);
        if !inside_safe_exponent {
            return am_error;
        }
    }
    let upperbit = (hi >> 63) as i32;
    let mut mantissa = hi >> (upperbit + 64 - F::MANTISSA_EXPLICIT_BITS as i32 - 3);
    let mut power2 = power(q as i32) + upperbit - lz as i32 - F::MINIMUM_EXPONENT;
    if power2 <= 0 {
        if -power2 + 1 >= 64 {
            return am_zero;
        }
        mantissa >>= -power2 + 1;
        mantissa += mantissa & 1;
        mantissa >>= 1;
        power2 = (mantissa >= (1_u64 << F::MANTISSA_EXPLICIT_BITS)) as i32;
        return AdjustedMantissa { mantissa, power2 };
    }
    if lo <= 1
        && q >= F::MIN_EXPONENT_ROUND_TO_EVEN as i64
        && q <= F::MAX_EXPONENT_ROUND_TO_EVEN as i64
        && mantissa & 3 == 1
        && (mantissa << (upperbit + 64 - F::MANTISSA_EXPLICIT_BITS as i32 - 3)) == hi
    {
        mantissa &= !1_u64;
    }
    mantissa += mantissa & 1;
    mantissa >>= 1;
    if mantissa >= (2_u64 << F::MANTISSA_EXPLICIT_BITS) {
        mantissa = 1_u64 << F::MANTISSA_EXPLICIT_BITS;
        power2 += 1;
    }
    mantissa &= !(1_u64 << F::MANTISSA_EXPLICIT_BITS);
    if power2 >= F::INFINITE_POWER {
        return am_inf;
    }
    AdjustedMantissa { mantissa, power2 }
}
```

**Comment by Alexhuszagh** [comment]
@lemire I've mostly been using the term "moderate" path for that, to disambiguate from the term "fast path" used in dec2flt. Since the above uses an extended-precision representation of the significant digits of the float, it would replace the Bellerophon algorithm, which is used if `fast_path` cannot be used. Rust's dec2flt has a function named `fast_path` which only works in cases where both the significant digits and the exponent can exactly be represented as native floats (see below). In my opinion, the whole thing should be thrown out and replaced with the above code sample, but I'm working on providing minimal changes with major performance tweaks in the meantime.

https://github.com/rust-lang/rust/blob/28e2b29b8952485679367cc05699fb5154f4e5c3/library/core/src/num/dec2flt/algorithm.rs#L109-L144

**Comment by Alexhuszagh** [comment]
Oh I should probably add: Rust's dec2flt doesn't parse the first 19-20 significant digits of the float except in the fast path, which is a major performance issue and trivial to fix (and then your algorithm could then be implemented into Rust's core library). The [fast path](https://github.com/rust-lang/rust/blob/28e2b29b8952485679367cc05699fb5154f4e5c3/library/core/src/num/dec2flt/algorithm.rs#L113-L144) just sees if there is less than 16-digits in the float's integral and fractional components, and then checks if these significant digits can be exactly represented as a native float.

If not, it defaults to [creating a big integer](https://github.com/rust-lang/rust/blob/28e2b29b8952485679367cc05699fb5154f4e5c3/library/core/src/num/dec2flt/mod.rs#L271-L277) from the significant digits, and then calls it's unoptimized Bellerophon algorithm, which can lead to [extreme](https://github.com/rust-lang/rust/issues/85234) performance hits for no good reason.

In addition, the subsequent [algorithm_r](https://github.com/rust-lang/rust/blob/28e2b29b8952485679367cc05699fb5154f4e5c3/library/core/src/num/dec2flt/algorithm.rs#L185-L241) ideally prefers a value of `z0` that is within 1 ULP of the correct value. This is trivial to implement, using your above algorithm, but would of course have to be modified for that purpose.

**Comment by lemire** [comment]
@Alexhuszagh 

Right. 

You have Clinger's fast path which is very cheap and handles things like (small) integers and so forth.

Otherwise, what you want to do is to grab, say, the first 19 digits. Most of the time, numbers won't have more than 17 digits. If you have no more than 19 digits, then the code above is good enough  99.99999% of the time, and even if you have more than 19 digits, the truncation is good enough almost all of the time. See section 11... https://arxiv.org/pdf/2101.11408.pdf

So it should be uncommon that you ever need to materialize a big integer.

**Comment by Alexhuszagh** [comment]
Yep. My [modifications](https://github.com/Alexhuszagh/rust-dec2flt/compare/infer...moderate) keep Clinger's algorithm (since I don't believe anyone currently on the Rust core team has worked on dec2flt before), but moves significant digit parsing outside of the fast path, and adds error calculation for truncated values, and only after that do we use their big-integer algorithms (which have major issues in nearly every facet, but that's a different story).

Ideally, we'd remove it, and replace it with your code, but I'm trying minor, incremental improvements at first. Ideally, my goals right now are:

1. Fix disguised fast path cases.
2. Improve the Bellerophon algorithm, to handle subnormal floats and avoid needlessly using big-integers.
3. Fix the glacially slow `algorithm_m` and `algorithm_r` with easy-to-implement slow-path algorithms.
4. Increase the size of the big-integer storage, so it can handle more than 385 digits (since we need up to 767, which means a lot of floats cannot be parsed by Rust).
5. Remove the last remaining Clinger algorithm, Bellerophon, and replace it with the Eisel-Lemire algorithm based on the code above.

**Comment by Urgau** [comment]
> Ideally, we'd remove it, and replace it with your code, but I'm trying minor, incremental improvements at first. Ideally, my goals right now are:
> 
>     1. Fix disguised fast path cases.
>
>     2. Improve the Bellerophon algorithm, to handle subnormal floats and avoid needlessly using big-integers.
>
>     3. Fix the glacially slow `algorithm_m` and `algorithm_r` with easy-to-implement slow-path algorithms.
>
>     4. Increase the size of the big-integer storage, so it can handle more than 385 digits
>        (since we need up to 767, which means a lot of floats cannot be parsed by Rust).
>
>     5. Remove the last remaining Clinger algorithm, Bellerophon, and replace it with the Eisel-Lemire algorithm
>        based on the code above.

*From an outside point of view:* Wouldn't be it simpler and faster to *simply* replace the actual rust code with the code of fast-float ?

As I see things, there is actual many faults in the rust std implementation, and you are proposing to do a 5 steps improvement (but not a total fix ?). I think that instead of trying to patch the std rust code to ideally be perfect we should take the fastest road and completely replace the actual suboptimal and broken (apparently) by a working one. In this case fast-float or another crate if it fits better the need of the std library.

**Comment by Alexhuszagh** [comment]
@Urgau Sure, if the maintainers are up for that. There's a few issues mentioned though, and that is:

1. The original implementer is no longer working on Rust, so no one is familiar with the codebase I believe.
2. There's a few assumptions internally that need to be addressed prior to making these changes.
3. The changes required here are relatively trivial, and therefore easy to justify, prior to making larger changes.

If you can justify the changes, that would be ideal. I'm going to test the binary sizes, but this crate should be pretty lean (likely leaner than core), actually. Also, sure it's not a total fix, but it's an improvement of like 10,000x for some floats, which is not insignificant.

**Comment by Alexhuszagh** [comment]
Ok I've run the numbers and the binary sizes are practically identical, albeit slightly smaller.

**fast-float**

opt-level|size|size(stripped)
|:-:|:-:|:-:|
0|3.3M|324K
1|3.3M|300K
2|1.3M|248K
3|1.3M|252K
s|1.3M|244K
z|1.3M|248K

**core**

opt-level|size|size(stripped)
|:-:|:-:|:-:|
0|3.6M|360K
1|3.5M|316K
2|1.3M|236K
3|1.3M|248K
s|1.3M|244K
z|1.3M|248K

If you can get the core team to approve of it, it would be a great addition to the core library. I'm still going to try to get incremental improvements implemented, for the reasons above (IMO, it's easier to justify), but ideally, this should be the end goal.

You would likely, however, have to use the [bignum[(https://github.com/rust-lang/rust/blob/master/library/core/src/num/bignum.rs) implementation for the Decimal struct.

As an FYI: I would recommend forking my dec2flt fork of Rust's `num` module if you're going to do this, since it allows you to implement these changes within the context of the Rust core library without having to build Rust from scratch every time. It's a minimal fork of the dec2flt library, in the context of the Rust core library, and should make it quick and easy to make changes, diffs, and then incorporate them back into the core library.

**Comment by Urgau** [comment]
> @Urgau Sure, if the maintainers are up for that.

Posted an [question](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Replacing.20dec2flt.20by.20fast-float.20in.20core) on the zulip stream for T-Libs.

> Ok I've run the numbers and the binary sizes are practically identical, albeit slightly smaller.

That great news.

> As an FYI: I would recommend forking my dec2flt fork of Rust's `num` module if you're going to do this, since it allows you to implement these changes within the context of the Rust core library without having to build Rust from scratch every time. It's a minimal fork of the dec2flt library, in the context of the Rust core library, and should make it quick and easy to make changes, diffs, and then incorporate them back into the core library.

Thanks for the tips but I'm not sure I'm qualified to try doing the integration in std. In fact I barely understand how float are store internally. I will read the academic paper but I'm not sure it will be sufficient.

**Comment by Urgau** [comment]
@Alexhuszagh I have receive an response from Josh Triplett (who is co-lead of the Rust Libs Team). He said [here](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Replacing.20dec2flt.20by.20fast-float.20in.20core/near/238623493): 

> [..] this seems like a great idea to me [...]

So if you want try to do the pull request replacing the dec2flt with the fast-float code, go for it.

There is however an legal concern also raised by Josh.
> fast-float-rust says it's directly based on the C++ library. Is the Rust implementation derived from the C++ code or the algorithm in the paper? Asking because the C++ code is just Apache-2.0, but the Rust code says MIT/Apache-2.0.

> If the Rust code were a derived work of the C++, it wouldn't be able to add the MIT dual license. (And the C++ code has had third-party contributions, so it isn't just the original author who would need to grant permission for that.)

@aldanor @lemire Would you mind clarifying for Josh Triplett the situation here or directly on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Replacing.20dec2flt.20by.20fast-float.20in.20core) concerning the licenses.

Also I have being acting like your were okay for your implementation to be part of the rust standard library without never asking you if you were okay for that. So do you agree for your implementation to be part of the rust standard library ?

**Comment by lemire** [comment]
I have zero ownership of anything. I am just an interested fellow.

**Comment by lemire** [comment]
If anyone wants to chat about the algorithm, I am available (via Zoom). I did not write a single code of Rust for this project though. I am just around to help.

**Comment by Alexhuszagh** [comment]
> Thanks for the tips but I'm not sure I'm qualified to try doing the integration in std. In fact I barely understand how float are store internally. I will read the academic paper but I'm not sure it will be sufficient.

I was more so responding to the author of the crate in this section, sorry if that wasn't clear.

**Comment by aldanor** [comment]
I'm also "just an interested fellow", in a way - as Daniel is the true owner of the algorithm, and I was porting the C++ implementation to Rust (with some minor implementational changes and updates here and there) while also having the paper by my side as a reference.

If there are any license concerns - we can sure resolve them; I'm pretty sure all of us are interested first and foremost in this to succeed. We can either make the original library Apache2/MIT (which would require all contributors to sign off on that if I understand it correctly), or make this crate Apache2-only. Not a big deal.

If Josh and the rest of the core team are happy to go ahead with this - should we give it a go? I'd be happy to have a separate zoom/zulip/whatever discussion if needed. Never had experience of contributing anything major to rustc myself, but as long as there's some guidance re: what needs to go where, I think I can try.

**Comment by Alexhuszagh** [comment]
@aldanor Rust is dual-licensed under both MIT/Apache-2.0, so I believe keeping the original license intact. I'm more than happy to have a Zoom call, if needed, or chat on Zulip. I've had a failed attempt (personal reasons forced me to take a step back) previously at integrating changes into the core library, but I'm tangentially familiar with the process. I would be more than happy to contribute any help I can.

**Comment by Alexhuszagh** [comment]
<s>Looking carefully at the implementation, there's likely a few things that would additionally need to be done to integrate this into Rust core.</s>
1. <s>Validate <a href="https://github.com/aldanor/fast-float-rust/blob/4680af7666df5b341c6bc46ab7a112ce1adea615/src/number.rs#L100-L118">digits</a> occur after the exponent symbol.</s>
2. <s>Ensure there are <a href=https://github.com/aldanor/fast-float-rust/blob/4680af7666df5b341c6bc46ab7a112ce1adea615/src/number.rs#L122-L166">significant digits</a> at all. Rust does not allow floats of the format `e10`, `.e10`, `\`, `+`, or `.`. You currently only handle `+` and `\`.</s>
3. <s><a href="https://github.com/aldanor/fast-float-rust/blob/4680af7666df5b341c6bc46ab7a112ce1adea615/src/number.rs#L231-L248">Case-sensitive</a> matches on special values.</s>

<s>A simple playground of all these examples is <a href="https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=50fe526250b5a698793a72f1b1f43b75">here</a>.</s>

**Comment by Urgau** [comment]
> Looking carefully at the implementation, there's likely a few things that would additionally need to be done to integrate this into Rust core.
> 
>  1. Validate [digits](https://github.com/aldanor/fast-float-rust/blob/4680af7666df5b341c6bc46ab7a112ce1adea615/src/number.rs#L100-L118) occur after the exponent symbol.
>   2. Ensure there are [significant digits](https://github.com/aldanor/fast-float-rust/blob/4680af7666df5b341c6bc46ab7a112ce1adea615/src/number.rs#L122-L166) at all. Rust does not allow floats of the format `e10`, `.e10`, `\`, `+`, or `.`. You currently only handle `+` and `\`.

These doesn't seems to be allowed by the IEEE 754 standard either.

> 3. [Case-sensitive](https://github.com/aldanor/fast-float-rust/blob/4680af7666df5b341c6bc46ab7a112ce1adea615/src/number.rs#L231-L248) matches on special values.

This was a bug on the implementation, that have been fix on nightly some weeks ago. See [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018) and also [here](https://github.com/rust-lang/rust/pull/78618).

**Comment by Alexhuszagh** [comment]
Ah cool, then only 1). and 2). are necessary.

**Comment by aldanor** [comment]
Hmm..

```rust
fn main() {
    for s in &[
        "e10", "1.23", "1.23e", "1.", ".1", "1.23e5", ".", ".e1", "0e", "0e1",
    ] {
        println!("{}", s);
        println!("  std: {:?}", s.parse::<f64>());
        println!("  ff:  {:?}", fast_float::parse::<f64, _>(s));
    }
}
```

yields (as of the current version)

```
e10
  std: Err(ParseFloatError { kind: Invalid })
  ff:  Err(Error)
1.23
  std: Ok(1.23)
  ff:  Ok(1.23)
1.23e
  std: Err(ParseFloatError { kind: Invalid })
  ff:  Err(Error)
1.
  std: Ok(1.0)
  ff:  Ok(1.0)
.1
  std: Ok(0.1)
  ff:  Ok(0.1)
1.23e5
  std: Ok(123000.0)
  ff:  Ok(123000.0)
.
  std: Err(ParseFloatError { kind: Invalid })
  ff:  Err(Error)
.e1
  std: Err(ParseFloatError { kind: Invalid })
  ff:  Err(Error)
0e
  std: Err(ParseFloatError { kind: Invalid })
  ff:  Err(Error)
0e1
  std: Ok(0.0)
  ff:  Ok(0.0)
```

**Comment by Alexhuszagh** [comment]
Nevermind, I missed this: `*s = start`. My fault. So then we look good to go.

**Comment by Urgau** [comment]
@aldanor I have [discuss](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Replacing.20dec2flt.20by.20fast-float.20in.20core/near/238730640) with Josh and the only way forward with license is to have fast_float be dual license under MIT/APACHE.

I have open a [pull request](https://github.com/fastfloat/fast_float/pull/70) in the fast_float repository to change the license.

> If Josh and the rest of the core team are happy to go ahead with this - should we give it a go? I'd be happy to have a separate zoom/zulip/whatever discussion if needed. Never had experience of contributing anything major to rustc myself, but as long as there's some guidance re: what needs to go where, I think I can try.

Yep, you should give it a try. You should also watch the [rustc-dev-guide](https://rustc-dev-guide.rust-lang.org/getting-started.html), there are some very useful information about compiling rustc (shouldn't be necessary) and the standard library.

The files that are relevant for you are under the [library/core/src/num/dev2flt](https://github.com/rust-lang/rust/tree/master/library/core/src/num/dec2flt) directories. From what I can see all of the files except `mod.rs` who needs to be modify can be deleted/replaced with your fast_float port.

**Comment by lemire** [comment]
+1

**Comment by Alexhuszagh** [comment]
Looking at the [commits](https://github.com/fastfloat/fast_float/commits) for fast-float, only [@biojppm](https://github.com/fastfloat/fast_float/commits?author=biojppm) and [@kitaisreal ](https://github.com/fastfloat/fast_float/commits?author=kitaisreal) made commits prior to this port being made, and so would likely be the only ones who would need to agree to re-license their code so an (older) version of fast-float could be dual licensed, and therefore this crate dual-licensed.

We could ask permission from both of them. kitaisreal's contributions are trivial, so we could likely remove those changes if licensing becomes a major issue.

**EDIT**: Their name is "kitaisreal".

**Comment by lemire** [comment]
I have emailed @biojppm

**Comment by Alexhuszagh** [comment]
I've got on my own fork all the data tests from Rust's core library, and I'll be running them tonight on a little-endian (32-bit) and big-endian (32-bit) architecture, since those are the only cases that haven't yet been comprehensively tested IIRC. None of the logic should affect these cases, but it's nice to be extra sure.

I'll be running this [branch](https://github.com/Alexhuszagh/fast-float-rust/tree/comprehensive) with all the unittests within for the following two architectures:
- armv7-unknown-linux-gnueabihf (32-bit, LE)
- powerpc-unknown-linux-gnu (32-bit, BE)

This is going to take a while, since these tests basically trigger the slow algorithm every time, but it shouldn't be too bad.

**Comment by Alexhuszagh** [comment]
We have correctness issues on PowerPC 32-bit, BE. I'm going to run the tests again, but here's a few examples:

```text
0.5000309541037827 ULP error on 1303e-20 (f64)
0.5000309541037827 ULP error on 2606e-20 (f64)
0.5000309541037827 ULP error on 5212e-20 (f64)
```

I'm going to check the cases to see if it could possibly be a soft-float emulation issue (these issues do happen, I had to open a few patches in Qemu before). I'll have some specialized tests going to triage this, and I'll get back to you.

`armv7-unknown-linux-gnueabihf` ran without issue. All of these are examples of fast-path cases (and not disguised fast-path cases either), so it's likely this is a soft float issue. A lot of stuff uses soft floats on PPC (due to the awful performance of hard floats on PPC makes soft floats faster), so this might not even be a Qemu issue, it might just be a PPC issue. A trivial way to check would be to try these same floats compiled for PPC as well.

**Comment by Alexhuszagh** [comment]
Ok I've triaged the issue.

I've added the following unnittests:

```rust
use fast_float::parse;

#[test]
fn test_f64_soft_float() {
    assert_eq!(
        parse::<f64, _>("1303e-20").unwrap(),
        f64::from_bits(0b11110001101110000010111000110111101101101111110001101110101000)
    );
    assert_eq!(
        parse::<f64, _>("2606e-20").unwrap(),
        f64::from_bits(0b11110001111110000010111000110111101101101111110001101110101000)
    );
    assert_eq!(
        parse::<f64, _>("5212e-20").unwrap(),
        f64::from_bits(0b11110010001110000010111000110111101101101111110001101110101000)
    );
}
```

And then compile analogous to this:

```bash
RUSTFLAGS='-C soft-float' cargo test
RUSTFLAGS='-C soft-float' cross test --target powerpc-unknown-linux-gnu
RUSTFLAGS='-C soft-float' cross test --target armv7-unknown-linux-gnueabihf
```

The following targets fail:
- powerpc64le-unknown-linux-gnu
- powerpc-unknown-linux-gnu

The following targets pass:
- armv7-unknown-linux-gnueabihf (with and without the `-C soft-float` option.
- i686-unknown-linux-gnu (with and without the `-C soft-float` option.
- mips-unknown-linux-gnu 

So it's not an endian issue, it seems to be a PowerPC issue. Looking at the bit patterns of the power (e-17), using the `pow10_fast_path` we get:

```rust
// Code
// println!("e17={:?}", (f64::pow10_fast_path(17)).to_bits());
// println!("e-17={:?}", (1.0/f64::pow10_fast_path(17)).to_bits());

// MIPS, ARMv7, etc.
e17=4861130398305394688
e-17=4352464011485697175

// PowerPC
e17=4861130398305394688
e-17=4352464011485697175
```

So the issue doesn't manifest in simple cases. When doing a little more poking, we get:

```
// Code
// println!("1.303e17={:?}", (1.303 * f64::pow10_fast_path(17)).to_bits());
// println!("1.303e-17={:?}", (1.303/f64::pow10_fast_path(17)).to_bits());

// MIPS, ARMv7, etc.
1.303e17=4863024148305394688
1.303e-17=4354430593920867240

// PowerPC
1.303e17=4863024148305394688
1.303e-17=4354430593920867240
```

Which is likewise identical. However, we get a 1 ULP error with the following:

```rust
// Code
// println!("1e20={:?}", (1.0 * f64::pow10_fast_path(20)).to_bits());
// println!("1e-20={:?}", (1.0/f64::pow10_fast_path(20)).to_bits());
// println!("1303e20={:?}", (1303.0 * f64::pow10_fast_path(20)).to_bits());
// println!("1303e-20={:?}", (1303.0/f64::pow10_fast_path(20)).to_bits());

// PowerPC
1e20=4906019910204099648
1e-20=4307583784117748259
1303e20=4952718876067038006
1303e-20=4354430593920867241
```

This seems to be an error in float division on PowerPC systems. I'm going to quickly check if other systems have similar issues.

**Comment by Alexhuszagh** [comment]
This seems to be a PPC bug or Qemu bug. Not too worried about it then. I used both `-mhard` in GCC and the `-C soft-float` in rustc and both produced the same error. It's either an artifact of Qemu, in which case, who cares. It could also be a bug in PPC, in which case, who cares. It's not an issue with the library.

```c
#include <inttypes.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>

int main() {
  double num = 1303.0;
  double den = 1e20;
  double value = num / den;
  uint64_t bits;
  memcpy(&bits, &value, 8);
  printf("%" PRIu64 "\n", bits);

  return 0;
}
```

```bash
# powerpc-linux-gnu-gcc main.c -o main -std=c99
# qemu-ppc ./main
4354430593920867241
```

So.... I think we're good to go... 🎉🎉

**Comment by Urgau** [comment]
> So.... I think we're good to go... 🎉🎉

Great! Happy to see this moving so fast.

**Comment by lemire** [comment]
I have an actual POWER9 processor. I can give you guys access, just email me at `lemire@gmail.com`.

I get ...

```
$ gcc main.c -o main -std=c99
$ ./main
4354430593920867240
```

This is an *actual* POWER9 system.

<img width="343" alt="Screen Shot 2021-05-15 at 10 30 12 AM" src="https://user-images.githubusercontent.com/391987/118364918-89ea6c00-b568-11eb-94e0-22dd3d1dd34a.png">

**Comment by lemire** [comment]
@Alexhuszagh I have an actual POWER9 machine. I can either run the tests for you if you give me the command lines you want me to run (ideally within a docker container) or else you can email me at lemire@gmail.com and I can give you ssh access.

**Comment by Alexhuszagh** [comment]
@lemire I think we're good then. Seems like an artifact. Can you create the assembly using:

```bash
gcc -S main.c
more main.s
```

That's all I really need to know. All the magic numbers seem to be correct, so I think this is just a Qemu bug (time for me to file another issue). I'd expect something with the following:

```asm
; The actual load, fdiv, and store instructions
; ....

; The magic numbers, this is what we'd expect. LC0 is 
; 1303.0, and LC1 is 1e20, just in two 32-bit words.

; 1303.0 == transmuting ((1083464704 << 32) | 0)
.LC0:
        .long   0          
        .long   1083464704 
        .align 8

; 1e20 == transmuting ((1142271773 << 32) | 2025163840)
.LC1:
        .long   2025163840 
        .long   1142271773
```

I'm generating that assembly, there doesn't seem to be a precision flag (there's nothing weird like the x87, this is all handled correctly by the `f*` instructions (`fdiv` for 64-bit, `fdivs` for 32-bit).

**x87 weirdness**
![Screenshot_20210515_092708](https://user-images.githubusercontent.com/9440903/118369523-50663080-b569-11eb-9cf7-1523a8d889ac.png)

**Comment by lemire** [comment]
I am not an expert in POWER assembly, but here is the dump. There is the fdiv instruction. It is hard to read, but if I compile just...

```
double divide(double num, double den) {
 return num / den;
}
```

I get effectively a single instruction...

```
fdiv 1,1,2
```


Note that I would be very surprised (and disappointed) if POWER processors did not follow the IEEE standard and common practices to the T. IBM takes standards seriously.


```
$ more main.s
	.file	"main.c"
	.abiversion 2
	.section	".text"
	.section	.rodata
	.align 3
.LC2:
	.string	"%lu\n"
	.section	".text"
	.align 2
	.globl main
	.type	main, @function
main:
.LFB0:
	.cfi_startproc
.LCF0:
0:	addis 2,12,.TOC.-.LCF0@ha
	addi 2,2,.TOC.-.LCF0@l
	.localentry	main,.-main
	mflr 0
	std 0,16(1)
	std 31,-8(1)
	stdu 1,-144(1)
	.cfi_def_cfa_offset 144
	.cfi_offset 65, 16
	.cfi_offset 31, -8
	mr 31,1
	.cfi_def_cfa_register 31
	addis 9,2,.LC0@toc@ha
	addi 9,9,.LC0@toc@l
	lfd 0,0(9)
	stfd 0,96(31)
	addis 9,2,.LC1@toc@ha
	addi 9,9,.LC1@toc@l
	lfd 0,0(9)
	stfd 0,104(31)
	lfd 12,96(31)
	lfd 0,104(31)
	fdiv 0,12,0
	stfd 0,112(31)
	ld 9,112(31)
	std 9,120(31)
	ld 9,120(31)
	mr 4,9
	addis 3,2,.LC2@toc@ha
	addi 3,3,.LC2@toc@l
	bl printf
	nop
	li 9,0
	mr 3,9
	addi 1,31,144
	.cfi_def_cfa 1, 0
	ld 0,16(1)
	mtlr 0
	ld 31,-8(1)
	blr
	.long 0
	.byte 0,0,0,1,128,1,0,1
	.cfi_endproc
.LFE0:
	.size	main,.-main
	.section	.rodata
	.align 3
.LC0:
	.long	0
	.long	1083464704
	.align 3
.LC1:
	.long	2025163840
	.long	1142271773
	.ident	"GCC: (Debian 8.3.0-6) 8.3.0"
	.section	.note.GNU-stack,"",@progbits
```

**Comment by Alexhuszagh** [comment]
@lemire The magic numbers (which is what's important, the rest is just effectively load, store, and division instructions) are all the same: this is a Qemu bug. Thanks for verifying. Further triaging, it seems this was fixed in Qemu-4.2.1, so this is specifically a cross issue that I've [filed](https://github.com/rust-embedded/cross/issues/557). 

As far as correctness goes, as long as we use the FPU [precision flags](https://github.com/Alexhuszagh/fast-float-rust/blob/comprehensive/src/fpu.rs) for x87 that Rust already has, we're good to go!

**Comment by Urgau** [comment]
My [pull request](https://github.com/fastfloat/fast_float/pull/70) for dual licensing [fast_float](https://github.com/fastfloat/fast_float/) has been merged. We're now also go to go with licensing. 🎉

**Comment by Alexhuszagh** [comment]
I've expanded my tests by the way to:
- powerpc-unknown-linux-gnu
- armv7-unknown-linux-gnueabihf
- x86_64-unknown-linux-gnu
- s390x-unknown-linux-gnu
- mips64el-unknown-linux-gnuabi64
- mips-unknown-linux-gnu
- arm-unknown-linux-gnueabihf
- i686-unknown-linux-gnu
- aarch64-unknown-linux-gnu
- powerpc-unknown-linux-gnu
- armv7-unknown-linux-gnueabihf
- x86_64-unknown-linux-gnu
- riscv64gc-unknown-linux-gnu

I think we can guarantee that we have complete correctness on every Tier 1 and Tier 2 target.

**Comment by aldanor** [comment]
Here's another concern I have: let's say we integrate this into core (and maybe, although doubtfully so at the moment, we even get `from_parts()` API in core).

What happens to `parse_partial()`? This is completely irreplaceable in many parsing scenarios, and that's really the main entry point here.

**Comment by Alexhuszagh** [comment]
@aldanor Pretty good reason to maintain the library, IMO. One of my realizations early on with lexical was that there was a number of use-cases that would never make it into core, so I think there's a good rationale for the continued existence of this library from a generalized parsing perspective. Still, numerous users don't use 3rd-party crates for a lot of functionality, whether by choice or simply due to a lack of awareness. The number of people who would therefore benefit from such inclusion would be massive.

**Comment by aldanor** [comment]
Yea, that's my thoughts too. Coexistence of two (and perhaps slightly different) versions of the same code to be maintained is not the most cheerful prospect of course... Although if `from_parts()` is accepted, then at least that part wouldn't have to be duplicated.

**Comment by Urgau** [comment]
@Alexhuszagh @aldanor I plan to write an update on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Replacing.20dec2flt.20by.20fast-float.20in.20core), but there have been no interaction is the past week; is there any update regarding the integration in core ?

Also, do you need help and/or mentoring ? If that's the case I can ask on zulip.

**Comment by Alexhuszagh** [comment]
@Urgau I've been doing some enhancements on this crate, fast_float (the C++ implementation), and my own crate lexical. I'm more than happy to spearhead the integration into core if that's ok @aldanor? I was going to defer to you, since it's your crate, but if you're fine with me taking the lead, I will gladly put in the work.

**Comment by aldanor** [comment]
@Urgau @Alexhuszagh I was planning to start looking into that in detail this weekend, unfortunately I didn't have time up to this point for personal reasons :/

@Alexhuszagh I won't mind of course, that sounds great, and you have more experience with num/dec2flt since you've already played around with it. I would be glad to participate in reviewing then.

Also: if there's any other enhancements you had in mind (like endian stuff etc), perhaps now is the best time to do as much of that as possible on this crate before integration into core so this crate and its future stdlib version would diverge as little as possible?

**Comment by Alexhuszagh** [comment]
@aldanor My last changes (I've been looking at both this crate and the C++ implementation extensively) are fixed by my open PRs. I have no further improves planned, except changes to add x87 support when integrating into core (which is impossible on a 3rd party crate).

So if all is good, I'm more than happy to spearhead this. My endian [changes](https://github.com/fastfloat/fast_float/pull/75) have been merged in [fast_float](https://github.com/fastfloat/fast_float), and once they're merged here, I believe we should be good to go.

**Comment by Urgau** [comment]
> @Urgau @Alexhuszagh I was planning to start looking into that in detail this weekend, unfortunately I didn't have time up to this point for personal reasons :/

No problem, take your time.

> @Alexhuszagh I won't mind of course, that sounds great, and you have more experience with num/dec2flt since you've already played around with it. I would be glad to participate in reviewing then.

This seems to be a good idea.

> Also: if there's any other enhancements you had in mind (like endian stuff etc), perhaps now is the best time to do as much of that as possible on this crate before integration into core so this crate and its future stdlib version would diverge as little as possible?

Should we consider having this crate as an subtree ? so that development can still continue here and do subtree update once in a while.

> So if all is good, I'm more than happy to spearhead this. My endian changes have been merged in fast_float, and once they're merged here, I believe we should be good to go.

That's great news. 😀

**Comment by Alexhuszagh** [comment]
I've set up a tracking branch then for the changes [here](https://github.com/Alexhuszagh/rust-dec2flt/tree/fast-float). I'll be using this to create a diff I'll then apply to rust-lang proper (to avoid re-compiling rust every time). There's going to have to be one significant change from the source code (which isn't actually that significant, in the grand scheme of things):
- AdjustedMantissa is effectively just `diy_float::Fp`, which is also used by Grisu. 
 
This is a pretty trivial fix, but we do need to be careful. The exponent in `diy_float::Fp` is not biased, while `AdjustedMantissa` is.

**Comment by lemire** [comment]
@Alexhuszagh Should you need help, ping me.

**Comment by Urgau** [comment]
@Alexhuszagh @lemire Any update on the integration ? I didn't see any pull request on the [rust-lang/rust](https://github.com/rust-lang/rust) repository.

**Comment by lemire** [comment]
I am still available to help but I have not done anything.

**Comment by Alexhuszagh** [comment]
@Urgau Sorry a perfect storm of vacation, Linux distro upgrades (my Fedora 32 went EOL, and that also meant an upgrade to Wayland, which resulting in a lot of bugs), and work all hit. I've got all the base functionality on my tree, and now am ensuring the changes ensure no backwards incompatible changes with the existing implementation (since `fast-float-rust` doesn't have very good error handling types, among a few other things).

**Comment by Alexhuszagh** [comment]
# Status

Ok I've got a fairly good initial draft [here](https://github.com/Alexhuszagh/rust-dec2flt/tree/fast-float). Notice there are a few differences:

1. More documentation (this is good, since the maintainers may not necessarily be us).
2. Changed large tables from using `const` to `static`. This is a known bug in the Rust compiler and can lead to much slower compile times.
3. Re-factored a bit (minimal changes as was possible) to simplify the following changes.
4. Added support for `ParseFloatError` with `ParseFloatErrorKind` to support both `Invalid` and `Empty`, to avoid backwards-incompatible changes.
5. Added `set_precision` for x87 FPUs to ensure no rounding errors occur on older x86 models.
6. Removed inlining. This is just a thing Rust core does. We'll see if they're open to re-adding it.

Let me know what you think: I'll be adding more documentation, soliciting feedback, and then submitting it for a PR.

# TODO

1. I need to re-enable some external tests that were previously internal.
2. Module-level documentation.
3. Better function-level documentation.
4. Finish the comprehensive tests and benchmarks.
5. Create a diff, apply it to `rust-lang/rust`, and submit a PR.

**Comment by Alexhuszagh** [comment]
# Benchmarks

Here are my results running on my x86_64 machine (same as before, more than willing to share exact specs if desired):

**dec2flt**

```text
dec2flt/fast            time:   [21.038 ns 21.145 ns 21.293 ns]                          
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  5 (5.00%) high severe
dec2flt/disguised       time:   [21.595 ns 21.740 ns 21.908 ns]                               
Found 8 outliers among 100 measurements (8.00%)
  5 (5.00%) high mild
  3 (3.00%) high severe
dec2flt/moderate        time:   [36.496 ns 36.775 ns 37.184 ns]                              
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) high mild
  8 (8.00%) high severe
dec2flt/halfway         time:   [36.215 ns 36.487 ns 36.751 ns]                             
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
dec2flt/large           time:   [34.656 us 34.928 us 35.261 us]                           
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe
dec2flt/denormal        time:   [76.883 ns 77.461 ns 78.156 ns]                             
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
```

**fast-float-rust**

```text
fast-float/fast         time:   [21.178 ns 21.355 ns 21.529 ns]                             
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
Benchmarking fast-float/disguised: Collecting 100 samples in estimated 5.0000 s (216M iterations                                                                                                fast-float/disguised    time:   [22.820 ns 23.071 ns 23.419 ns]
Found 14 outliers among 100 measurements (14.00%)
  5 (5.00%) high mild
  9 (9.00%) high severe
fast-float/moderate     time:   [37.724 ns 37.821 ns 37.935 ns]                                 
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
fast-float/halfway      time:   [40.177 ns 40.682 ns 41.219 ns]                                
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
fast-float/large        time:   [17.482 us 17.579 us 17.685 us]                              
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe
fast-float/denormal     time:   [73.703 ns 76.554 ns 79.831 ns]                                
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe
```
As you can see, the results are practically identical, except for `large` values.

**Comment by Urgau** [comment]
> Ok I've got a fairly good initial draft [here](https://github.com/Alexhuszagh/rust-dec2flt/tree/fast-float). Notice there are a few differences:
> 
>     1. More documentation (this is good, since the maintainers may not necessarily be us).

This is always a good things to have accurate and good documentation.

>     2. Changed large tables from using `const` to `static`. This is a known bug in the Rust compiler
>        and can lead to much slower compile times.

Yeah, that's sad.

>     4. Added support for `ParseFloatError` with `ParseFloatErrorKind` to support both `Invalid` and `Empty`,
>        to avoid backwards-incompatible changes.

👍

>     5. Added `set_precision` for x87 FPUs to ensure no rounding errors occur on older x86 models.

Does the actual rust std library also do that ?

>     6. Removed inlining. This is just a thing Rust core does. We'll see if they're open to re-adding it.

This doesn't seems to be truth, I found several use of `#[inline]` in core and std. As far as I know only `#[inline(always)]` is prity-much always refuse useless proof that it significantly helps.

> As you can see, the results are practically identical, except for large values.

That's great. 😀

Do you know why the large benchmark has regressed so much ?

**Comment by Alexhuszagh** [comment]
> Does the actual rust std library also do that ?

Yep. Although it may not matter in the wild (since all the values can be exactly represented in the number of significant digits represented by the type, and the default behavior is the round to an 80-bit, extended-precision float defined by the x87, which can cause exactness issues. In our case, it's unlikely to be an issue unless the `Precision Control field` was set to `0x0000` for a 32-bit float, rather than `0x3000` (80-bit) or `0x2000` (64-bit), since this would **definitely** cause rounding issues. You can see the original implementation [here](https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/library/core/src/num/dec2flt/algorithm.rs#L31-L109).

You can see the actual documentation for this [here](https://www.felixcloutier.com/x86/fstcw:fnstcw) and [here](http://www.nacad.ufrj.br/online/intel/Documentation/en_US/compiler_f/main_for/fpops/fortran/fpops_contw_f.htm).

> This doesn't seems to be truth, I found several use of `#[inline]` in core and std. As far as I know only `#[inline(always)]` is prity-much always refuse useless proof that it significantly helps.

It depends where. For most type primitives, it's done, but often simple traits don't even do it. See [AsRef](https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/library/core/src/convert/mod.rs#L596-L608) for a good example of this. More complex functions, such as [core/slice/index.rs::range](https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/library/core/src/slice/index.rs#L516) and [core::ops::RangeBounds::contains](https://github.com/rust-lang/rust/blob/a50d72158e08e02cfc051b863017bdbd2c45b637/library/core/src/ops/range.rs#L810) never seem to inline.

Since this all seems to be the latter, rather than the former, I've removed almost all the inlining and am more than happy to add it back based on feedback.

> Do you know why the large benchmark has regressed so much ?

Guessing inlining, but I'll have to check exactly why. None of the changes besides inlining would explain the difference. It's also possible it was a single benchmark abnormality, I'm planning on running it a bit more. Please note that these are engineered to be rare cases, so for most real data sets, it won't matter that much.

**Comment by Alexhuszagh** [comment]
Good comparisons of inlining for similar functionality can be found in [core/num/diy_float.rs](https://github.com/rust-lang/rust/blob/75d1500f028d0e5dcb3978b873bf2d0a0be9c75f/library/core/src/num/diy_float.rs), [core/num/bignum.rs](https://github.com/rust-lang/rust/blob/75d1500f028d0e5dcb3978b873bf2d0a0be9c75f/library/core/src/num/bignum.rs), [core/num/flt2dec/strategy/dragon.rs](https://github.com/rust-lang/rust/blob/75d1500f028d0e5dcb3978b873bf2d0a0be9c75f/library/core/src/num/flt2dec/strategy/dragon.rs), and [core/num/flt2dec/strategy/grisu.rs](https://github.com/rust-lang/rust/blob/75d1500f028d0e5dcb3978b873bf2d0a0be9c75f/library/core/src/num/flt2dec/strategy/grisu.rs)

As you can see, there's not a single instance of inlining in **any** of them, so I'd be relatively cautious to make major departures from this.

**Comment by Alexhuszagh** [comment]
@aldanor This is undefined behavior, and must be replaced:

https://github.com/aldanor/fast-float-rust/blob/ec1b7d4ff98dc177b1f507e53c0f39a611542811/src/common.rs#L76-L79

[CERT C Reference](https://wiki.sei.cmu.edu/confluence/display/c/ARR36-C.+Do+not+subtract+or+compare+two+pointers+that+do+not+refer+to+the+same+array)

[ptr::add](https://doc.rust-lang.org/std/primitive.pointer.html#safety-5)

[LLVM Reference](https://llvm.org/docs/LangRef.html#addresses-of-basic-blocks)

**Comment by Alexhuszagh** [comment]
Ok I've just added an initial version of the commit for the [PR](https://github.com/Alexhuszagh/rust) I will be submitted, and am currently building it and modifying the test suite accordingly, based on an up-to-date master branch of Rust.

Things currently left to do:
- Add a script to automatically generate the constants in `table.rs`. This should replace the deleted [dec2flt_table.py](https://github.com/Alexhuszagh/rust/blob/6e0b554619a3bb7e75b3334e97f191af20ef5d76/src/etc/dec2flt_table.py).
- Ensure all the tests using publicly exported internal modules pass (`dec2flt/parse.rs` and `dec2flt/rawfp.rs`).
- Add more tests to replace the existing tests.

I've added pretty extensive internal documentation for everything for everything to hopefully facilitate any work of future maintainers, and feedback would be great.

**Comment by lemire** [comment]
For `decimal.rs`, I suggest you credit Nigal Tao and Ken Thompson. The name of the technique according to Rob Pike is "Simple Decimal Conversion".

Regarding `compute_float`, it looks like 'magic' and that's not a good thing. I suggest
you provide the formal peer-reviewed reference which describes the algorithm in detail,
along with the mathematical derivations:

> Daniel Lemire, Number Parsing at a Gigabyte per Second, Software: Practice and Experience (to appear) https://arxiv.org/abs/2101.11408

For the script, the script I used is presented in the paper cited above and found online at
https://github.com/fastfloat/fast_float/blob/main/script/table_generation.py

**Comment by Alexhuszagh** [comment]
This should all be done as of https://github.com/Alexhuszagh/rust/commit/24f5414d795729c8d5167beef40e790b2abda4d6. Let me know if all the changes seem satisfactory, and I'll submit the PR today.

**Comment by lemire** [comment]
@Alexhuszagh Very nice.

**Comment by Alexhuszagh** [comment]
A PR has been [submitted](https://github.com/rust-lang/rust/pull/86761).

**Comment by Urgau** [comment]
@Alexhuszagh  @lemire I saw on zulip this PR https://github.com/rust-lang/rust/issues/31407 about valid float being treated as invalid. Will the new algorithm fix this ?

**Comment by lemire** [comment]
@Urgau  Nice point.

**Comment by Alexhuszagh** [comment]
@Urgau That is actually addressed in the PR, in the introduction. Not the issue specifically, but the general idea:

> Finally, this algorithm provides substantial improvements in the number of floats the Rust core library can parse. Denormal floats with a large number of digits cannot be parsed, due to use of the `Big32x40`, which simply does not have enough digits to round a float correctly. Using a custom decimal class, with much simpler logic, we can parse all valid decimal strings of any digit count.

```rust
// Issue in Rust's dec2fly.
"2.47032822920623272088284396434110686182e-324".parse::<f64>();   // Err(ParseFloatError { kind: Invalid })
```

So that's absolutely part of the motivation.

**Comment by Alexhuszagh** [comment]
Major updates on the PR:

1. Almost all the unsafe code has been removed, without affecting performance.
2. A lot of internal compiler error workarounds have been solved, giving even more motivation for the PR.
3. A lot of code has been de-duplicated, reducing the PR size substantially.
4. All comments on code logic so far seem to have been addressed, and the code seems to be simple enough to clearly audit. A lot of preconditions have been documented, and a lot of "complex" sections of the code have been extensively documented so they're easy to understand.

The only major bottleneck right now seems to be they're waiting for a computer scientist who is familiar with float-parsing to comment on the PR. I think I know an expert....

**Comment by Urgau** [comment]
@Alexhuszagh This is great. Thank you for creating the PR and taking it to the end.

I just have a small suggestion, it might be nice to add/update the benchmarks comparing speed and code size between the current implementation and the PR one.

PS: I just noticed that this PR https://github.com/rust-lang/rust/pull/86048 is going to have conflicts with yours, you will certainly have to be rebased.

**Comment by Alexhuszagh** [comment]
> @Alexhuszagh This is great. Thank you for creating the PR and taking it to the end.
> 
> I just have a small suggestion, it might be nice to add/update the benchmarks comparing speed and code size between the current implementation and the PR one.

I'll update that after most of the changes have been made: there's one more involving Miri that I'm finishing. I doubt the benchmarks have changed much, because I've ensured every change to ensure the safe routines don't change the resulting ASM. See this repository for all the [changes](https://github.com/Alexhuszagh/fast-float-asm). In fact, everything besides `step_by` is not just practically (which I've kept unsafe, since it would have significant performance penalties), but actually identical, 

> PS: I just noticed that this PR [rust-lang/rust#86048](https://github.com/rust-lang/rust/pull/86048) is going to have conflicts with yours, you will certainly have to be rebased.

Yeah, once there's conflicting changes, I'll be more than happy to rebase. However, as it's not yet merged, I'll wait. I'll likely do a rebase once the Miri issues are fixed as well.

**Comment by lemire** [comment]
@Alexhuszagh This code looks great, let me know if there is anything specific that needs to be examined.

**Comment by Urgau** [comment]
The pull-request for integrating the algorthim in the rust standard library has been [merged](https://github.com/rust-lang/rust/pull/86761#issuecomment-881914858) 🎉.

Thank you @Alexhuszagh for having created and brought this pull-request to the end.
Thank's also to you @lemire and @aldanor for your help.

**Comment by lemire** [comment]
@Urgau @Alexhuszagh Great work. One fun challenge would be to re-use the same tables for serialization. It is on my todo to investigate this option.

**Comment by Alexhuszagh** [comment]
@lemire Thanks for all the hard work, and helpful feedback. As of right now... 🎉🎉🎉 :crab rave:

**Comment by lemire** [comment]
@Alexhuszagh Do not hesitate to get in touch if you'd like to continue this work and you feel I can help.
